### ISSUE TITLE
Nickel Formatting

### ISSUE THREAD

**Issue by vkleen** [tweag/topiary#331]
I tried formatting some code from the Nickel standard library using topiary. First of all, it's really close to being awesome 🎉 The primary issues I see are:
- Empty lines being deleted between long record fields
- Type signature and documentation metadata being made to start on the same line
- Documentation closing delimiter and field definition being put on the same line

I don't actually know how to change the queries to address, so I decided to open this issue to at least report my experience.

Here is an excerpt:
```nickel
{
  function = {
    id : forall a. a -> a
    | doc m%"
    The identity function, that is
    ```nickel
    id x == x
    ```
    for any value `x`.
    "%
    = fun x => x,

    const : forall a b. a -> b -> a
    | doc m%"
    Results in the first argument.

    For example:
    ```nickel
    const 5 42 == 5
    ```
    "%
    = fun x y => x,

    compose : forall a b c. (b -> c) -> (a -> b) -> (a -> c)
    | doc m%"
    Function composition, right to left. That is,
    ```nickel
    compose f g x == f (g x)
    ```
    "%
    = fun g f x => x |> f |> g,
  }
}
```
Formatting this results in
```nickel
{
  function = {
    id : forall a. a -> a | doc m%"
    The identity function, that is
    ```nickel
    id x == x
    ```
    for any value `x`.
    "% = fun x => x,
    const : forall a b. a -> b -> a | doc m%"
    Results in the first argument.

    For example:
    ```nickel
    const 5 42 == 5
    ```
    "% = fun x y => x,
    compose : forall a b c. (b -> c) -> (a -> b) -> (a -> c) | doc m%"
    Function composition, right to left. That is,
    ```nickel
    compose f g x == f (g x)
    ```
    "% = fun g f x => x |> f |> g,
  }
}
```

**Comment by Xophmeister** [comment]
I've been playing with this, today. Let me document my progress so far:

>> * Empty lines being deleted between long record fields

Sorted :+1:

>> * Type signature and documentation metadata being made to start on the same line
> * If there is one metadata, on the same line as the definition, then let everything on the same line (I think Topiary has this distinction one-line/multiline in general, is that correct?)

Topiary does make a single vs. multi-line distinction, but it doesn't make a single vs. multiple children distinction. (At least not easily; it _may_ be possible with some query magic, but I'm not convinced... I'll see what I can do :slightly_smiling_face:) Anyway, for sake of argument, let's assume we can't make that distinction. Thus what we can achieve with Topiary is something like this:

  ```nickel
  # Single-line
  let a : TYPE | DOC | CONTRACT = 123

  # Multi-line
  let foo
    : TYPE
    | DOC
    | CONTRACT
    = 123
  ```

The trigger for multi-line is the input, so if the original annotations are spread over several lines, Topiary will respect that.

> * if a defintion has more than one metadata attached, IMHO the more readable and consistent way is to have one line per metadata definition (at least) and indent them

As precedent, the Nickel standard library keeps type annotations on the same line, but flows `|` annotations each on to their own line. Either way is easy :slightly_smiling_face: 

  ```nickel
  # All annotations on a new line (in multi-line context)
  let foo
    : TYPE
    | DOC
    | CONTRACT
    = 123

  # VS: Keep the type annotation together
  let foo : TYPE
    | DOC
    | CONTRACT
    = 123
  ```

> I think it's nice to ident doc content two times, and delimiter one.

The doc content is parsed as a string; the internals of which are not formatted. (Nickel's indent-aware multiline strings are quite unique.) As such, I think the only option would be that it's up to the author of the code to align the content appropriately.

> Another solution is to put the identifier and the first metadata on the same line:
> ```
> some_long_stuff_field | SomeContract -> SomeOtherContract
>                       | optional
>                       | rec default,
> short_one | String
>           | not_exported,
> ```

Topiary cannot do this (...yet: see #170)

**Comment by vkleen** [comment]
Would it be possible to keep the type annotation together up to a character or complexity limit? If not, I think putting all annotations on a new line would be the most reasonable.

**Comment by yannham** [comment]
Ah, your distinction of one-line versus multiline makes actually more sense, and was in fact what I meant :+1: 

> As precedent, the Nickel standard library keeps type annotations on the same line, but flows | annotations each on to their own line. Either way is easy

>Would it be possible to keep the type annotation together up to a character or complexity limit? If not, I think putting all annotations on a new line would be the most reasonable.

Honestly, the precedent set by the stdlib is mostly by accident. Some functions may have started with only a type annotation, and then were enriched as we implemented new metadata in Nickel. I personally don't see an excellent reason why type annotations should be treated differently than other metadata. I'm in favor of just putting the type annotation on a new line, like:

```nickel
# All annotations on a new line (in multi-line context)
let foo
  : TYPE
  | DOC
  | CONTRACT
  = 123
```

> Another solution is to put the identifier and the first metadata on the same line:

> Topiary cannot do this (...yet: see https://github.com/tweag/topiary/issues/170)

Well, I'm fine with that, as I dislike this option :stuck_out_tongue: but wanted to mention it for the sake of completeness or if someone has a convincing motivation for it.

**Comment by yannham** [comment]
Another point I forgot to mention: it is currently customary to write default values like this:

```nickel
{
  SomeContract = {
    value
      | doc "hello"
      | optional
      | not_exported
      | default = 5,
  },
}
```

As if we set a default value. However, semantically, `default` is actually just a normal metadata annotation (a priority annotation) and `5` is the definition of this field. So, a formattnig that doesn't do anything special about default could end up changing that to:

```nickel
{
  SomeContract = {
    value
      | doc "hello"
      | optional
      | not_exported
      | default
      = 5,
  },
}
```

I'm not yet sure the current tradition is the right way, but if it were, would the former (1st) example possible to produce with current topiary? That is, if there's a default annotation among the children, put it last and immediately followed by the `= <def>`, that is not adding a newline between the two?  (then if `def` is large, we would of course split it on several lines as any code block).

**Comment by Xophmeister** [comment]
@vkleen 
> Would it be possible to keep the type annotation together up to a character or complexity limit?

Alas, no.

@yannham 
> I'm not yet sure the current tradition is the right way, but if it were, would the former (1st) example possible to produce with current topiary? That is, if there's a default annotation among the children, put it last and immediately followed by the = <def>, that is not adding a newline between the two? (then if def is large, we would of course split it on several lines as any code block).

The children can't be reordered, but it ought to be possible to match on `| default =`.

**Comment by Xophmeister** [comment]
Resolved by #167, #331, #353, #378 and #385
